Enables TRT-LLM backend to be used for target_verify

pranavm-nvidia opened the pull request:
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.ai to discuss further. -->

## Motivation

This change allows the TRT-LLM MLA backend to be used for the `target_verify` step in MTP, which should improve performance. 

## Modifications

- Enables `forward_extend` in the TRT-LLM MLA backend to be used for `target_verify` in MTP.
- Also adds code to update the KV cache in `forward_extend` which was previously missing.

## Accuracy Tests

### GSM8k:

<details>

Server Command:
```bash
python -m sglang.launch_server --model-path "deepseek-ai/DeepSeek-R1" --trust-remote-code --attention-backend trtllm_mla --page-size 64 --tp-size 8 --max-running-requests 512 --cuda-graph-max-bs 512 --host "0.0.0.0" --port 30000 --mem-fraction-static 0.70 --dist-timeout 1800 --speculative-algorithm EAGLE --speculative-eagle-topk=1 --speculative-num-draft-tokens 4 --speculative-num-steps 3
```

Test:
```
$ python3 benchmark/gsm8k/bench_sglang.py --num-shots 5 --num-questions 1319 --parallel 512 --port 30000

Accuracy: 0.946
Invalid: 0.000
Latency: 25.739 s
Output throughput: 5051.383 token/s
```
</details>

### GPQA-Diamond:

<details>

**Note: Please see [my comment below](https://github.com/sgl-project/sglang/pull/10281#issuecomment-3277067897) regarding the accuracy with MTP=off; it doesn't seem like MTP=on causes an accuracy drop compared to that.**

Server Command:
```bash
python -m sglang.launch_server --model-path "deepseek-ai/DeepSeek-R1" --trust-remote-code --attention-backend trtllm_mla --page-size 64 --tp-size 8 --max-running-requests 512 --cuda-graph-max-bs 512 --host "0.0.0.0" --port 30000 --mem-fraction-static 0.70 --dist-timeout 1800 --speculative-algorithm EAGLE --speculative-eagle-topk=1 --speculative-num-draft-tokens 4 --speculative-num-steps 3
```

Results:
| Category                                   | evaluation_mode | num_entries | avg_tokens | gen_seconds | symbolic_correct | no_answer |
|---------------------------------------------|-----------------|-------------|------------|-------------|------------------|-----------|
| gpqa                                       | pass@1          | 198         | 5414       | 329         | 70.20%           | 0.00%     |
| gpqa-Physics (general)                      | pass@1          | 19          | 4118       | 329         | 78.95%           | 0.00%     |
| gpqa-Organic Chemistry                      | pass@1          | 72          | 7983       | 326         | 44.44%           | 0.00%     |
| gpqa-Quantum Mechanics                      | pass@1          | 25          | 3521       | 266         | 100.00%          | 0.00%     |
| gpqa-Electromagnetism and Photonics         | pass@1          | 6           | 2103       | 74          | 83.33%           | 0.00%     |
| gpqa-High-energy particle physics           | pass@1          | 14          | 3768       | 299         | 92.86%           | 0.00%     |
| gpqa-Genetics                               | pass@1          | 4           | 4562       | 162         | 75.00%           | 0.00%     |
| gpqa-Astrophysics                           | pass@1          | 13          | 4758       | 215         | 92.31%           | 0.00%     |
| gpqa-Molecular Biology                      | pass@1          | 15          | 3711       | 306         | 80.00%           | 0.00%     |
| gpqa-Chemistry (general)                    | pass@1          | 20          | 5138       | 302         | 65.00%           | 0.00%     |
| gpqa-Relativistic Mechanics                 | pass@1          | 7           | 2545       | 113         | 85.71%           | 0.00%     |
| gpqa-Inorganic Chemistry                    | pass@1          | 1           | 6055       | 165         | 100.00%          | 0.00%     |
| gpqa-Optics and Acoustics                   | pass@1          | 1           | 2147       | 60          | 100.00%          | 0.00%     |
| gpqa-Condensed Matter Physics               | pass@1          | 1           | 1042       | 32          | 100.00%          | 0.00%     |


</details>

### Math-500

<details>

Server Command:
```bash
python -m sglang.launch_server --model-path "deepseek-ai/DeepSeek-R1" --trust-remote-code --attention-backend trtllm_mla --page-size 64 --tp-size 8 --max-running-requests 512 --cuda-graph-max-bs 512 --host "0.0.0.0" --port 30000 --mem-fraction-static 0.70 --dist-timeout 1800 --speculative-algorithm EAGLE --speculative-eagle-topk=1 --speculative-num-draft-tokens 4 --speculative-num-steps 3
```

Results:
| evaluation_mode | num_entries | avg_tokens | gen_seconds | symbolic_correct | no_answer |
|-----------------|-------------|------------|-------------|------------------|-----------|
| pass@1          | 500         | 2284       | 370         | 94.60%           | 1.80%     |

</details>

## Benchmarking and Profiling

### FlashInfer vs. TRT-LLM MLA (MTP=on for both)

<details>

Server commands:
```
python -m sglang.launch_server --model-path "deepseek-ai/DeepSeek-R1" --trust-remote-code --attention-backend flashinfer --page-size 64 --tp-size 8 --max-running-requests 512 --cuda-graph-max-bs 512 --host "0.0.0.0" --port 30000 --mem-fraction-static 0.70 --dist-timeout 1800 --speculative-algorithm EAGLE --speculative-eagle-topk=1 --speculative-num-draft-tokens 4 --speculative-num-steps 3

python -m sglang.launch_server --model-path "deepseek-ai/DeepSeek-R1" --trust-remote-code --attention-backend trtllm_mla --page-size 64 --tp-size 8 --max-running-requests 512 --cuda-graph-max-bs 512 --host "0.0.0.0" --port 30000 --mem-fraction-static 0.70 --dist-timeout 1800 --speculative-algorithm EAGLE --speculative-eagle-topk=1 --speculative-num-draft-tokens 4 --speculative-num-steps 3
```

Client:
```
python3 -m sglang.bench_serving --backend sglang --model "deepseek-ai/DeepSeek-R1" --num-prompts 256 --dataset-name random --random-input-len 1024 --random-output-len 1024 --random-range-ratio 1 --max-concurrency=256
```

Results:
| Metric                                    | FlashInfer Result | TRT-LLM Result | % Difference (TRT-LLM vs FlashInfer) |
|-------------------------------------------|:-----------------:|:--------------:|:------------------------------------:|
| Total input tokens                        | 262144            | 262144         | 0.00%                                |
| Total generated tokens                    | 262144            | 262144         | 0.00%                                |
| Total generated tokens (retokenized)      | 261347            | 261314         | -0.01%                               |
| Request throughput (req/s)                | 5.47              | 5.79           | +5.84%                               |
| Input token throughput (tok/s)            | 5598.17           | 5934.04        | +6.00%                               |
| Output token throughput (tok/s)           | 5598.17           | 5934.04        | +6.00%                               |
| Total token throughput (tok/s)            | 11196.34          | 11868.09       | +6.00%                               |
| Concurrency                               | 225.74            | 222.14         | -1.59%                               |
| Accept length                             | 2.62              | 2.79           | +6.49%                               |
| Mean E2E Latency (ms)                     | 41290.86          | 38332.63       | -7.23%                               |
| Median E2E Latency (ms)                   | 42046.48          | 39026.21       | -7.19%                               |
| Mean TTFT (ms)                            | 4382.98           | 4294.74        | -2.01%                               |
| Median TTFT (ms)                          | 4384.86           | 4308.44        | -1.75%                               |
| P99 TTFT (ms)                             | 8171.60           | 7987.98        | -2.25%                               |
| Mean ITL (ms)                             | 36.08             | 33.27          | -7.80%                               |
| Median ITL (ms)                           | 28.33             | 24.54          | -13.37%                              |
| P95 ITL (ms)                              | 45.87             | 46.65          | +1.70%                               |
| P99 ITL (ms)                              | 89.37             | 88.16          | -1.35%                               |
| Max ITL (ms)                              | 7688.61           | 7519.31        | -2.20%                               |


</details>

### TRT-LLM MLA MTP on vs. off (concurrency=1)

<details>

Server commands:
```
python -m sglang.launch_server --model-path "deepseek-ai/DeepSeek-R1" --trust-remote-code --attention-backend trtllm_mla --page-size 64 --tp-size 8 --max-running-requests 512 --cuda-graph-max-bs 512 --host "0.0.0.0" --port 30000 --mem-fraction-static 0.70 --dist-timeout 1800

python -m sglang.launch_server --model-path "deepseek-ai/DeepSeek-R1" --trust-remote-code --attention-backend trtllm_mla --page-size 64 --tp-size 8 --max-running-requests 512 --cuda-graph-max-bs 512 --host "0.0.0.0" --port 30000 --mem-fraction-static 0.70 --dist-timeout 1800 --speculative-algorithm EAGLE --speculative-eagle-topk=1 --speculative-num-draft-tokens 4 --speculative-num-steps 3
```

Client:
```
python3 -m sglang.bench_serving --backend sglang --model "deepseek-ai/DeepSeek-R1" --num-prompts 64 --dataset-name random --random-input-len 1024 --random-output-len 1024 --random-range-ratio 1 --max-concurrency=1
```

Results:
| Metric                                   | MTP=off           | MTP=on            | % Difference                |
|-------------------------------------------|-------------------|-------------------|-----------------------------|
| Total input tokens                        | 65536             | 65536             | 0%                          |
| Total generated tokens                    | 65536             | 65536             | 0%                          |
| Total generated tokens (retokenized)      | 65349             | 65372             | +0.04%                      |
| Request throughput (req/s)                | 0.08              | 0.15              | +87.50%                     |
| Input token throughput (tok/s)            | 82.72             | 151.53            | +83.23%                     |
| Output token throughput (tok/s)           | 82.72             | 151.53            | +83.23%                     |
| Total token throughput (tok/s)            | 165.44            | 303.07            | +83.32%                     |
| Concurrency                              | 1.00              | 1.00              | 0%                          |
| Accept length                            | -                 | 2.69              | N/A                         |
| Mean E2E Latency (ms)                    | 12378.26          | 6757.24           | -45.39%                     |
| Median E2E Latency (ms)                  | 12378.47          | 6738.44           | -45.54%                     |
| Mean TTFT (ms)                           | 102.32            | 97.48             | -4.73%                      |
| Median TTFT (ms)                         | 101.83            | 97.37             | -4.38%                      |
| P99 TTFT (ms)                            | 105.88            | 101.64            | -4.01%                      |
| Mean ITL (ms)                            | 12.00             | 6.51              | -45.75%                     |
| Median ITL (ms)                          | 12.00             | 5.77              | -51.92%                     |
| P95 ITL (ms)                             | 12.41             | 17.08             | +37.68%                     |
| P99 ITL (ms)                             | 12.55             | 17.78             | +41.78%                     |
| Max ITL (ms)                             | 16.64             | 21.41             | +28.70%                     |

</details>


### TRT-LLM MLA MTP on vs. off (concurrency=8)

<details>

Server commands:
```
python -m sglang.launch_server --model-path "deepseek-ai/DeepSeek-R1" --trust-remote-code --attention-backend trtllm_mla --page-size 64 --tp-size 8 --max-running-requests 512 --cuda-graph-max-bs 512 --host "0.0.0.0" --port 30000 --mem-fraction-static 0.70 --dist-timeout 1800

python -m sglang.launch_server --model-path "deepseek-ai/DeepSeek-R1" --trust-remote-code --attention-backend trtllm_mla --page-size 64 --tp-size 8 --max-running-requests 512 --cuda-graph-max-bs 512 --host "0.0.0.0" --port 30000 --mem-fraction-static 0.70 --dist-timeout 1800 --speculative-algorithm EAGLE --speculative-eagle-topk=1 --speculative-num-draft-tokens 4 --speculative-num-steps 3
```

Client:
```
python3 -m sglang.bench_serving --backend sglang --model "deepseek-ai/DeepSeek-R1" --num-prompts 64 --dataset-name random --random-input-len 1024 --random-output-len 1024 --random-range-ratio 1 --max-concurrency=8
```

Results:
| Metric                                   | MTP=off         | MTP=on          | % Difference                |
|-------------------------------------------|-----------------|-----------------|-----------------------------|
| Total input tokens                        | 65536           | 65536           | 0%                          |
| Total generated tokens                    | 65536           | 65536           | 0%                          |
| Total generated tokens (retokenized)      | 65319           | 65361           | +0.06%                      |
| Request throughput (req/s)                | 0.53            | 0.71            | +34.0%                      |
| Input token throughput (tok/s)            | 547.30          | 729.13          | +33.2%                      |
| Output token throughput (tok/s)           | 547.30          | 729.13          | +33.2%                      |
| Total token throughput (tok/s)            | 1094.61         | 1458.27         | +33.3%                      |
| Concurrency                              | 8.00            | 7.70            | -3.75%                      |
| Accept length                            | -               | 2.71            | N/A                         |
| Mean E2E Latency (ms)                     | 14964.81        | 10810.43        | -27.7%                      |
| Median E2E Latency (ms)                   | 14961.21        | 10750.30        | -28.1%                      |
| Mean TTFT (ms)                            | 240.68          | 143.85          | -40.2%                      |
| Median TTFT (ms)                          | 193.33          | 117.34          | -39.3%                      |
| P99 TTFT (ms)                             | 438.72          | 323.26          | -26.3%                      |
| Mean ITL (ms)                             | 14.39           | 10.43           | -27.5%                      |
| Median ITL (ms)                           | 14.36           | 8.96            | -37.6%                      |
| P95 ITL (ms)                              | 14.89           | 27.12           | +82.2%                      |
| P99 ITL (ms)                              | 15.05           | 30.33           | +101.6%                     |
| Max ITL (ms)                              | 260.28          | 478.39          | +83.8%                      |

</details>


### TRT-LLM MLA MTP on vs. off (concurrency=16)

<details>

Server commands:
```
python -m sglang.launch_server --model-path "deepseek-ai/DeepSeek-R1" --trust-remote-code --attention-backend trtllm_mla --page-size 64 --tp-size 8 --max-running-requests 512 --cuda-graph-max-bs 512 --host "0.0.0.0" --port 30000 --mem-fraction-static 0.70 --dist-timeout 1800

python -m sglang.launch_server --model-path "deepseek-ai/DeepSeek-R1" --trust-remote-code --attention-backend trtllm_mla --page-size 64 --tp-size 8 --max-running-requests 512 --cuda-graph-max-bs 512 --host "0.0.0.0" --port 30000 --mem-fraction-static 0.70 --dist-timeout 1800 --speculative-algorithm EAGLE --speculative-eagle-topk=1 --speculative-num-draft-tokens 4 --speculative-num-steps 3
```

Client:
```
python3 -m sglang.bench_serving --backend sglang --model "deepseek-ai/DeepSeek-R1" --num-prompts 64 --dataset-name random --random-input-len 1024 --random-output-len 1024 --random-range-ratio 1 --max-concurrency=16
```

Results:
| Metric                                   | MTP=off         | MTP=on          | % Difference         |
|-------------------------------------------|-----------------|-----------------|----------------------|
| Total input tokens                        | 65536           | 65536           | 0%                   |
| Total generated tokens                    | 65536           | 65536           | 0%                   |
| Total generated tokens (retokenized)      | 65319           | 65356           | +0.06%               |
| Request throughput (req/s)                | 0.87            | 1.13            | +29.89%              |
| Input token throughput (tok/s)            | 894.61          | 1157.60         | +29.41%              |
| Output token throughput (tok/s)           | 894.61          | 1157.60         | +29.41%              |
| Total token throughput (tok/s)            | 1789.22         | 2315.20         | +29.37%              |
| Concurrency                              | 15.99           | 14.71           | -8.00%               |
| Accept length                            | —               | 2.71            | —                    |
| Mean E2E Latency (ms)                     | 18307.75        | 13013.49        | -28.89%              |
| Median E2E Latency (ms)                   | 18289.69        | 13071.94        | -28.48%              |
| Mean TTFT (ms)                            | 287.83          | 179.31          | -37.70%              |
| Median TTFT (ms)                          | 285.44          | 125.67          | -55.98%              |
| P99 TTFT (ms)                             | 375.73          | 317.55          | -15.52%              |
| Mean ITL (ms)                             | 17.61           | 12.55           | -28.71%              |
| Median ITL (ms)                           | 17.61           | 10.33           | -41.33%              |
| P95 ITL (ms)                              | 18.05           | 31.29           | +73.38%              |
| P99 ITL (ms)                              | 18.21           | 53.58           | +194.29%             |
| Max ITL (ms)                              | 207.73          | 685.71          | +230.10%             |

</details>



### TRT-LLM MLA MTP on vs. off (concurrency=256)

<details>

Server commands:
```
python -m sglang.launch_server --model-path "deepseek-ai/DeepSeek-R1" --trust-remote-code --attention-backend trtllm_mla --page-size 64 --tp-size 8 --max-running-requests 512 --cuda-graph-max-bs 512 --host "0.0.0.0" --port 30000 --mem-fraction-static 0.70 --dist-timeout 1800

python -m sglang.launch_server --model-path "deepseek-ai/DeepSeek-R1" --trust-remote-code --attention-backend trtllm_mla --page-size 64 --tp-size 8 --max-running-requests 512 --cuda-graph-max-bs 512 --host "0.0.0.0" --port 30000 --mem-fraction-static 0.70 --dist-timeout 1800 --speculative-algorithm EAGLE --speculative-eagle-topk=1 --speculative-num-draft-tokens 4 --speculative-num-steps 3
```

Client:
```
python3 -m sglang.bench_serving --backend sglang --model "deepseek-ai/DeepSeek-R1" --num-prompts 256 --dataset-name random --random-input-len 1024 --random-output-len 1024 --random-range-ratio 1 --max-concurrency=256
```

Results:
| Metric                                   | MTP=off        | MTP=on         | % Difference         |
|-------------------------------------------|----------------|----------------|----------------------|
| Total input tokens                        | 262144         | 262144         | 0.00%                |
| Total generated tokens                    | 262144         | 262144         | 0.00%                |
| Total generated tokens (retokenized)      | 261253         | 261314         | +0.02%               |
| Request throughput (req/s)                | 5.33           | 5.79           | +8.62%               |
| Input token throughput (tok/s)            | 5461.20        | 5934.04        | +8.67%               |
| Output token throughput (tok/s)           | 5461.20        | 5934.04        | +8.67%               |
| Total token throughput (tok/s)            | 10922.40       | 11868.09       | +8.65%               |
| Concurrency                              | 255.80         | 222.14         | -13.17%              |
| Accept length                            | —              | 2.79           | —                    |
| Mean E2E Latency (ms)                    | 47962.98       | 38332.63       | -20.10%              |
| Median E2E Latency (ms)                  | 47963.28       | 39026.21       | -18.62%              |
| Mean TTFT (ms)                           | 4908.56        | 4294.74        | -12.50%              |
| Median TTFT (ms)                         | 4936.11        | 4308.44        | -12.71%              |
| P99 TTFT (ms)                            | 8532.59        | 7987.98        | -6.38%               |
| Mean ITL (ms)                            | 42.09          | 33.27          | -20.93%              |
| Median ITL (ms)                          | 38.45          | 24.54          | -36.17%              |
| P95 ITL (ms)                             | 39.91          | 46.65          | +16.90%              |
| P99 ITL (ms)                             | 40.35          | 88.16          | +118.52%             |
| Max ITL (ms)                             | 7936.68        | 7519.31        | -5.24%               |

</details>

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

fzyzcjy reviewed (commented):
hi, could you please run (1) math500 w/ 64k gen 3 times (3) gpqa diamond w/ 32k gen for at least 16 times (preferably 100 times) and paste the acc. I have seen cases when gsm8k is good but these benchs drop acc.

cc @kaixih - could you please share your modification to my script which helps run these

fzyzcjy commented:
looking forward to this PR, if it is hard to do the testing maybe merge first and I will test it in my case and report bugs if acc drops

zhyncs commented:
I still encountered this issue https://github.com/sgl-project/sglang/pull/10281#issuecomment-3290359853 @pranavm-nvidia

zhyncs commented:
```
python3 -m sglang.launch_server \
--model-path nvidia/DeepSeek-V3-0324-FP4 \
--tp 4 --attention-backend trtllm_mla \
--moe-runner-backend flashinfer_trtllm \
--quantization modelopt_fp4 \
--speculative-algorithm EAGLE --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4
```

```
python3 -m sglang.launch_server \
--model-path nvidia/DeepSeek-V3-0324-FP4 \
--tp 4 --attention-backend trtllm_mla \
--moe-runner-backend flashinfer_trtllm \
--quantization modelopt_fp4 \
--speculative-algorithm EAGLE --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4 \
--prefill-attention-backend fa4 --speculative-attention-mode decode
```

We need to ensure that this PR does not break the above two commands. Thanks! @pranavm-nvidia @kushanam

Commit:
Enables TRT-LLM backend to be used for target_verify

Enables `forward_extend` in the TRT-LLM MLA backend to be used
for `target_verify` in MTP.

Also adds code to update the KV cache in `forward_extend` which was
previously missing.

zhyncs commented:
The latest one still doesn't work.

```
python3 -m sglang.launch_server \
--model-path nvidia/DeepSeek-V3-0324-FP4 \
--tp 4 --attention-backend trtllm_mla \
--moe-runner-backend flashinfer_trtllm \
--quantization modelopt_fp4 \
--speculative-algorithm EAGLE --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4
```

```
  File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1784, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 2351, in forward
    hidden_states, residual = layer(
                              ^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1784, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 2080, in forward
    hidden_states = self.self_attn(
                    ^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1784, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 1190, in forward
    return self.forward_core(s)
           ^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 1254, in forward_core
    return self.forward_absorb_core(*inner_state)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 1472, in forward_absorb_core
    attn_output = self.attn_mqa(
                  ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1784, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/layers/radix_attention.py", line 108, in forward
    return forward_batch.attn_backend.forward(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/layers/attention/base_attn_backend.py", line 81, in forward
    return self.forward_extend(
           ^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/layers/attention/trtllm_mla_backend.py", line 671, in forward_extend
    workspace_buffer=metadata.workspace,
                     ^^^^^^^^^^^^^^^^^^
AttributeError: 'TRTLLMMLADecodeMetadata' object has no attribute 'workspace'
```

zhyncs commented:
gsm8k is lower than expected
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     8|exact_match|↑  |0.9158|±  |0.0076|
|     |       |strict-match    |     8|exact_match|↑  |0.8052|±  |0.0109|
```

```
python3 -m sglang.launch_server \
--model-path nvidia/DeepSeek-V3-0324-FP4 \
--tp 4 --attention-backend trtllm_mla \
--moe-runner-backend flashinfer_trtllm \
--quantization modelopt_fp4 \
--speculative-algorithm EAGLE --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4


pip3 install "lm_eval[api]"
lm_eval --model local-chat-completions --model_args model=gpt-oss,base_url=http://127.0.0.1:30000/v1/chat/completions,num_concurrent=128,timeout=999999,max_gen_toks=2048 --tasks gsm8k --batch_size 128 --apply_chat_template --num_fewshot 8
```

Commit:
working

Commit:
cleanup

Commit:
fixes

Commit:
removes draft_token_num

zhyncs committed:
Merge branch 'main' into trtllm-target-verify

fzyzcjy commented:
my fix is WIP in https://github.com/fzyzcjy/sglang/tree/feat/ac6280

fzyzcjy committed:
cherry pick fix

fzyzcjy committed:
Merge branch 'main' into trtllm-target-verify

fzyzcjy commented:
### Support of fp8

test command

```
SGLANG_USE_CUTLASS_BACKEND_FOR_FP4_GEMM=1 python3 -m sglang.launch_server \
--model-path /data/numa0/tom/downloaded_models/models--nvidia--DeepSeek-V3-0324-FP4/snapshots/d03662cdc34b56eab1315d4557e395e6b4944782 \
--tp 4 --attention-backend trtllm_mla \
--moe-runner-backend flashinfer_trtllm \
--quantization modelopt_fp4 \
--speculative-algorithm EAGLE --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4 \
--kv-cache-dtype fp8_e4m3
```

results (strict-match in lm-eval)

* 94.3, 94.8, 94.5

thus seems not bad

fzyzcjy committed:
cherry pick support fp8

fzyzcjy committed:
Merge branch 'trtllm-target-verify' of https://github.com/pranavm-nvidia/sglang into trtllm-target-verify

zhyncs committed:
Merge branch 'main' into trtllm-target-verify

zhyncs committed:
Merge branch 'main' into trtllm-target-verify

pavanimajety commented on python/sglang/srt/layers/attention/trtllm_mla_backend.py:
```diff
@@ -553,84 +571,134 @@ def forward_extend(
+        v = v.view(-1, layer.tp_k_head_num, layer.v_head_dim)
+
+        if forward_batch.forward_mode.is_target_verify():
+            metadata = (
```
Note for technical debt: We can refactor the change to this function to be reused between `forward_decode` and `target_verify`. We can simply pass in `ForwardMode` and modify the `seq_lens` and `q` tensors.

Fridge003 commented:
This PR causes some break when `--disable-chunked-prefix-cache` is added. 
Now trying to fix
